### PR TITLE
Add `.persist` annotation to opt into more efficient Dedalus persistence

### PR DESCRIPTION
**Pull request by shadaj**
Add `.persist` annotation to opt into more efficient Dedalus persistence

### CODE REVIEW

**Comment by shadaj** [comment]
Blocked on #546

File: hydroflow/src/scheduled/graph.rs [hunk 1]
```diff
@@ -282,11 +282,26 @@ impl Hydroflow {
...
             let sg_data = &self.subgraphs[sg_id.0];
             events_has_external |= is_external;
+
             if !sg_data.is_scheduled.replace(true) {
                 self.stratum_queues[sg_data.stratum].push_back(sg_id);
                 enqueued_count += 1;
             }
         }
+
+        if events_has_external {
```
Comment (zzlk) [inline hydroflow/src/scheduled/graph.rs:292]: remove this when #550 goes in

File: hydroflow_datalog_core/src/join_plan.rs [hunk 2]
```diff
@@ -409,12 +422,24 @@ pub fn expand_join_plan(
...
+                if left_expanded.persisted && !right_expanded.persisted && !is_anti {
+                    (parse_quote!('static), parse_quote!('tick), false)
+                } else if !left_expanded.persisted && right_expanded.persisted && !is_anti {
+                    (parse_quote!('tick), parse_quote!('static), false)
+                } else if left_expanded.persisted && right_expanded.persisted && !is_anti {
+                    (parse_quote!('static), parse_quote!('static), true)
+                } else {
+                    (parse_quote!('tick), parse_quote!('tick), false)
+                };
+
```
Comment (MingweiSamuel) [inline hydroflow_datalog_core/src/join_plan.rs:435]: I suppose eventually we'll want the hydroflow compiler to optimize these sorts of things a lot more which means less work here maybe.

Comment (MingweiSamuel) [reply hydroflow_datalog_core/src/join_plan.rs:435]: might be cleaner as a match
```rust
match (left_expanded.persisted, right_expanded.persisted, is_anti) {
    (true, false, false) => (parse_quote!('static), parse_quote!('tick), false),
    (false, true, false) => (parse_quote!('tick), parse_quote!('static), false),
    (true, true, false) => (parse_quote!('static), parse_quote!('static), true),
    _ => (parse_quote!('tick), parse_quote!('tick), false)
}
```
assume it compiles to the same thing probably

**Review by MingweiSamuel — APPROVED** [review]
What exactly does `.persist x` do? Can you say its equivalent to some sort of `x(...) :- x(...)` rule?

Kinda afraid the datalog code is getting a bit messy.

**Comment by shadaj** [comment]
Yeah, so `.persist x` opts into a "faster" version of `x(...) :+ x(...)`, that can use `'static` joins rather than reifying every tick. This is effectively hard-coding the join persistence e-graph rewrites, but because that is far from being ready this is a temporary measure.

**Pull request merged**